### PR TITLE
CR3 decoder cleanup and fixes

### PR DESCRIPTION
**Pull request by Floessie**
Hi there,

The new CR3 decoder ported over from LibRaw and then ART is of mixed code quality. This PR is meant to improve that step by step.

Problems identified so far:
- Silenced warnings
- Clang warnings
- Signed vs. unsigned
- `long` vs. `int32_t` and friends
- (Missing) endianess awareness
- `#defines`
- Mixed style

I hope we can work this out together.

Best,
Flössie

### CODE REVIEW

**Comment by heckflosse** [comment]
@Floessie Should we create an issue about the missing endian awareness at libraw, so they can fix their code?

**Comment by Floessie** [comment]
@heckflosse You're right. I'll try to get in contact...

**Comment by Floessie** [comment]
[Done.](https://github.com/LibRaw/LibRaw/issues/254)

**Comment by heckflosse** [comment]
@Floessie On Windows I now get this garbage:
grafik

This fixes it. Can you verify that it still works on a big endian machine?
```diff
diff --git a/rtengine/canon_cr3_decoder.cc b/rtengine/canon_cr3_decoder.cc
index f10a7c083..0ffaf93cc 100644
--- a/rtengine/canon_cr3_decoder.cc
+++ b/rtengine/canon_cr3_decoder.cc
@@ -625,7 +625,7 @@ libraw_inline void _BitScanReverse(DWORD *Index, unsigned long Mask)
 }
 uint32_t _byteswap_ulong(uint32_t x)
 {
-    #if BYTE_ORDER == BIG_ENDIAN
+    #if defined(BYTE_ORDER) && BYTE_ORDER == BIG_ENDIAN
         return x;
     #else
         return __builtin_bswap32(x);

```

**Comment by heckflosse** [comment]
@Floessie Or using the same as in dcraw.h also works here

```diff
diff --git a/rtengine/canon_cr3_decoder.cc b/rtengine/canon_cr3_decoder.cc
index f10a7c083..c6f7ee0d3 100644
--- a/rtengine/canon_cr3_decoder.cc
+++ b/rtengine/canon_cr3_decoder.cc
@@ -625,7 +625,7 @@ libraw_inline void _BitScanReverse(DWORD *Index, unsigned long Mask)
 }
 uint32_t _byteswap_ulong(uint32_t x)
 {
-    #if BYTE_ORDER == BIG_ENDIAN
+    #if __BYTE_ORDER__==__ORDER_BIG_ENDIAN__
         return x;
     #else
         return __builtin_bswap32(x);

```

**Comment by Floessie** [comment]
@heckflosse This is exactly like it looked on PPC to no surprise. :smile: 

> Or using the same as in dcraw.h also works here

I'll test this variant...

**Comment by Floessie** [comment]
Works. :+1: I'll push...

**Comment by heckflosse** [comment]
@Floessie I also ran cppcheck
```
[H:/rt58_rc/rtengine/canon_cr3_decoder.cc:1964] (style) Redundant initialization for 'lineBufL2'. The initialized value is overwritten before it is read. [redundantInitialization]
[H:/rt58_rc/rtengine/canon_cr3_decoder.cc:1964] (style) Redundant initialization for 'lineBufL2'. The initialized value is overwritten before it is read. [redundantInitialization]
[H:/rt58_rc/rtengine/canon_cr3_decoder.cc:2650] (style) Local variable 'curSubband' shadows outer variable [shadowVariable]
[H:/rt58_rc/rtengine/canon_cr3_decoder.cc:2650] (style) Local variable 'curSubband' shadows outer variable [shadowVariable]
[H:/rt58_rc/rtengine/canon_cr3_decoder.cc:2370] (error) Memory leak: paramBuf [memleak]
[H:/rt58_rc/rtengine/canon_cr3_decoder.cc:2347] (style) C-style pointer casting detected. C++ offers four different kinds of casts as replacements: static_cast, const_cast, dynamic_cast and reinterpret_cast. A C-style cast could evaluate to any of those automatically, thus it is considered safer if the programmer explicitly states which kind of cast is expected. See also: https://www.securecoding.cert.org/confluence/display/cplusplus/EXP05-CPP.+Do+not+use+C-style+casts. [cstyleCast]
[H:/rt58_rc/rtengine/canon_cr3_decoder.cc:2430] (style) C-style pointer casting detected. C++ offers four different kinds of casts as replacements: static_cast, const_cast, dynamic_cast and reinterpret_cast. A C-style cast could evaluate to any of those automatically, thus it is considered safer if the programmer explicitly states which kind of cast is expected. See also: https://www.securecoding.cert.org/confluence/display/cplusplus/EXP05-CPP.+Do+not+use+C-style+casts. [cstyleCast]
[H:/rt58_rc/rtengine/canon_cr3_decoder.cc:2509] (style) C-style pointer casting detected. C++ offers four different kinds of casts as replacements: static_cast, const_cast, dynamic_cast and reinterpret_cast. A C-style cast could evaluate to any of those automatically, thus it is considered safer if the programmer explicitly states which kind of cast is expected. See also: https://www.securecoding.cert.org/confluence/display/cplusplus/EXP05-CPP.+Do+not+use+C-style+casts. [cstyleCast]
[H:/rt58_rc/rtengine/canon_cr3_decoder.cc:2693] (style) C-style pointer casting detected. C++ offers four different kinds of casts as replacements: static_cast, const_cast, dynamic_cast and reinterpret_cast. A C-style cast could evaluate to any of those automatically, thus it is considered safer if the programmer explicitly states which kind of cast is expected. See also: https://www.securecoding.cert.org/confluence/display/cplusplus/EXP05-CPP.+Do+not+use+C-style+casts. [cstyleCast]
[H:/rt58_rc/rtengine/canon_cr3_decoder.cc:2702] (style) C-style pointer casting detected. C++ offers four different kinds of casts as replacements: static_cast, const_cast, dynamic_cast and reinterpret_cast. A C-style cast could evaluate to any of those automatically, thus it is considered safer if the programmer explicitly states which kind of cast is expected. See also: https://www.securecoding.cert.org/confluence/display/cplusplus/EXP05-CPP.+Do+not+use+C-style+casts. [cstyleCast]
[H:/rt58_rc/rtengine/canon_cr3_decoder.cc:2703] (style) C-style pointer casting detected. C++ offers four different kinds of casts as replacements: static_cast, const_cast, dynamic_cast and reinterpret_cast. A C-style cast could evaluate to any of those automatically, thus it is considered safer if the programmer explicitly states which kind of cast is expected. See also: https://www.securecoding.cert.org/confluence/display/cplusplus/EXP05-CPP.+Do+not+use+C-style+casts. [cstyleCast]
[H:/rt58_rc/rtengine/canon_cr3_decoder.cc:2985] (style) C-style pointer casting detected. C++ offers four different kinds of casts as replacements: static_cast, const_cast, dynamic_cast and reinterpret_cast. A C-style cast could evaluate to any of those automatically, thus it is considered safer if the programmer explicitly states which kind of cast is expected. See also: https://www.securecoding.cert.org/confluence/display/cplusplus/EXP05-CPP.+Do+not+use+C-style+casts. [cstyleCast]
[H:/rt58_rc/rtengine/canon_cr3_decoder.cc:1964] (style) Redundant initialization for 'lineBufL2'. The initialized value is overwritten before it is read. [redundantInitialization]
[H:/rt58_rc/rtengine/canon_cr3_decoder.cc:818] (style) The scope of the variable 'nextData' can be reduced. Warning: Be careful when fixing this message, especially when there are inner loops. Here is an example where cppcheck will write that the scope for 'i' can be reduced:
void f(int x)
{
    int i = 0;
    if (x) {
        // it's safe to move 'int i = 0;' here
        for (int n = 0; n < 10; ++n) {
            // it is possible but not safe to move 'int i = 0;' here
            do_something(&i);
        }
    }
}
When you see this message it is always safe to reduce the variable scope 1 level. [variableScope]
[H:/rt58_rc/rtengine/canon_cr3_decoder.cc:869] (style) The scope of the variable 'nextWord' can be reduced. Warning: Be careful when fixing this message, especially when there are inner loops. Here is an example where cppcheck will write that the scope for 'i' can be reduced:
void f(int x)
{
    int i = 0;
    if (x) {
        // it's safe to move 'int i = 0;' here
        for (int n = 0; n < 10; ++n) {
            // it is possible but not safe to move 'int i = 0;' here
            do_something(&i);
        }
    }
}
When you see this message it is always safe to reduce the variable scope 1 level. [variableScope]
[H:/rt58_rc/rtengine/canon_cr3_decoder.cc:870] (style) The scope of the variable 'nextByte' can be reduced. Warning: Be careful when fixing this message, especially when there are inner loops. Here is an example where cppcheck will write that the scope for 'i' can be reduced:
void f(int x)
{
    int i = 0;
    if (x) {
        // it's safe to move 'int i = 0;' here
        for (int n = 0; n < 10; ++n) {
            // it is possible but not safe to move 'int i = 0;' here
            do_something(&i);
        }
    }
}
When you see this message it is always safe to reduce the variable scope 1 level. [variableScope]
[H:/rt58_rc/rtengine/canon_cr3_decoder.cc:972] (style) The scope of the variable 'nSyms' can be reduced. Warning: Be careful when fixing this message, especially when there are inner loops. Here is an example where cppcheck will write that the scope for 'i' can be reduced:
void f(int x)
{
    int i = 0;
    if (x) {
        // it's safe to move 'int i = 0;' here
        for (int n = 0; n < 10; ++n) {
            // it is possible but not safe to move 'int i = 0;' here
            do_something(&i);
        }
    }
}
When you see this message it is always safe to reduce the variable scope 1 level. [variableScope]
[H:/rt58_rc/rtengine/canon_cr3_decoder.cc:1289] (style) The scope of the variable 'nSyms' can be reduced. Warning: Be careful when fixing this message, especially when there are inner loops. Here is an example where cppcheck will write that the scope for 'i' can be reduced:
void f(int x)
{
    int i = 0;
    if (x) {
        // it's safe to move 'int i = 0;' here
        for (int n = 0; n < 10; ++n) {
            // it is possible but not safe to move 'int i = 0;' here
            do_something(&i);
        }
    }
}
When you see this message it is always safe to reduce the variable scope 1 level. [variableScope]
[H:/rt58_rc/rtengine/canon_cr3_decoder.cc:2380] (style) The scope of the variable 'transformWidth' can be reduced. Warning: Be careful when fixing this message, especially when there are inner loops. Here is an example where cppcheck will write that the scope for 'i' can be reduced:
void f(int x)
{
    int i = 0;
    if (x) {
        // it's safe to move 'int i = 0;' here
        for (int n = 0; n < 10; ++n) {
            // it is possible but not safe to move 'int i = 0;' here
            do_something(&i);
        }
    }
}
When you see this message it is always safe to reduce the variable scope 1 level. [variableScope]
[H:/rt58_rc/rtengine/canon_cr3_decoder.cc:1688] (style) The unsigned expression 'subband->width' will never be negative so it is either pointless or an error to check if it is. [unsignedLessThanZero]
[H:/rt58_rc/rtengine/canon_cr3_decoder.cc:2650] (style) Local variable 'curSubband' shadows outer variable [shadowVariable]
[H:/rt58_rc/rtengine/canon_cr3_decoder.cc:2667] (style) Suspicious calculation. Please use parentheses to clarify the code. The code ''a&b?c:d'' should be written as either ''(a&b)?c:d'' or ''a&(b?c:d)''. [clarifyCalculation]
[H:/rt58_rc/rtengine/canon_cr3_decoder.cc:816] (style) Variable 'bitData' is assigned a value that is never used. [unreadVariable]
[H:/rt58_rc/rtengine/canon_cr3_decoder.cc:2380] (style) Variable 'transformWidth' is assigned a value that is never used. [unreadVariable]
[H:/rt58_rc/rtengine/canon_cr3_decoder.cc:2647] (style) Variable 'curSubband' is assigned a value that is never used. [unreadVariable]
canon_cr3_decoder.cc
```

File: rtengine/canon_cr3_decoder.cc [hunk 1]
```diff
@@ -450,8 +468,8 @@ int DCraw::parseCR3(unsigned long long oAtomList,
       }
 #define current_track RT_canon_CR3_data.crx_header[nTrack]
 
-      ImageWidth = get2();
-      ImageHeight = get2();
+//      ImageWidth = get2();
```
Comment (heckflosse) [inline rtengine/canon_cr3_decoder.cc]: Couldn't it be possible the get2() are still needed because they change the fileposition?

Comment (Floessie) [reply rtengine/canon_cr3_decoder.cc]: You're right, though it still decoded...

**Comment by heckflosse** [comment]
I still get this warnings
```
Z:/H2/rt58_rc/rtengine/canon_cr3_decoder.cc: In member function 'int DCraw::parseCR3(long long unsigned int, long long unsigned int, short int&, char*, short unsigned int&, short int&)':
Z:/H2/rt58_rc/rtengine/canon_cr3_decoder.cc:539:41: warning: comparison of integer expressions of different signedness: 'long int' and 'uint32_t' {aka 'unsigned int'} [-Wsign-compare]
  539 |                 while (relpos_inDir + 6 < RT_canon_CR3_data.crx_header[nTrack].MediaSize) {
      |                        ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Z:/H2/rt58_rc/rtengine/canon_cr3_decoder.cc:556:49: warning: comparison of integer expressions of different signedness: 'long int' and 'long unsigned int' [-Wsign-compare]
  556 |                         while (relpos_inBox + 8 < relpos_inDir + szItem) {
      |                                ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~
```

Using -Wextra I get this additional warnings:
```
Z:/H2/rt58_rc/rtengine/canon_cr3_decoder.cc: In function 'int {anonymous}::crxReadSubbandHeaders({anonymous}::crx_data_header_t*, {anonymous}::CrxImage*, {anonymous}::CrxTile*, {anonymous}::CrxPlaneComp*, uint8_t**, uint32_t*)':
Z:/H2/rt58_rc/rtengine/canon_cr3_decoder.cc:2668:46: warning: unused parameter 'hdr' [-Wunused-parameter]
 2668 | int crxReadSubbandHeaders(crx_data_header_t* hdr, CrxImage* img, CrxTile* tile,
      |                           ~~~~~~~~~~~~~~~~~~~^~~
Z:/H2/rt58_rc/rtengine/canon_cr3_decoder.cc: In function 'int {anonymous}::crxSetupImageData({anonymous}::crx_data_header_t*, {anonymous}::CrxImage*, int16_t*, uint64_t, uint32_t, uint8_t*)':
Z:/H2/rt58_rc/rtengine/canon_cr3_decoder.cc:2983:23: warning: comparison is always false due to limited range of data type [-Wtype-limits]
 2983 |     if (img->tileCols > 0xFF || img->tileRows > 0xFF ||
      |         ~~~~~~~~~~~~~~^~~~~~
Z:/H2/rt58_rc/rtengine/canon_cr3_decoder.cc:2983:47: warning: comparison is always false due to limited range of data type [-Wtype-limits]
 2983 |     if (img->tileCols > 0xFF || img->tileRows > 0xFF ||
      |                                 ~~~~~~~~~~~~~~^~~~~~
```

**Comment by Floessie** [comment]
@heckflosse I think I'm done. Would be nice to get rid of the raw buffers, but I'd rather have this merged for 5.8 as is. Could you take a look and merge if everything is fine?

**Comment by heckflosse** [comment]
@Floessie Ok to merge with my latest changes?

**Comment by Floessie** [comment]
Wait a minute, your commit made me aware of other `bool` opportunities... I'll merge after that.

**Pull request merged**